### PR TITLE
ExportedNodeResolver - Add support for attributes

### PR DESCRIPTION
**Pull request by olsavmic**
Resolves https://github.com/phpstan/phpstan/issues/6797, https://github.com/phpstan/phpstan/issues/7413

### CODE REVIEW

**Comment by olsavmic** [comment]
@ondrejmirtes Could you give me a hint on how to test this properly? I don't see any test cases related to `ExportedNode*` or `DependencyResolver` in the codebase.

Thank you!

File: src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+use JsonSerializable;
+use PHPStan\Dependency\ExportedNode;
+
+class ExportedAttributeArgumentNode implements ExportedNode, JsonSerializable
+{
+
+	public function __construct(
+		private ?string $name,
+		private string $value,
+		private bool $byRef,
```
Comment (olsavmic) [inline src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php]: TODO: Most likely not even possible, test it

Comment (ondrejmirtes) [reply src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php]: I don't think we need the name here, being indexed correctly by `int|string` in ExportedAttributeNode should be fine.

File: src/Dependency/ExportedNode/ExportedAttributeNode.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+use PHPStan\Dependency\ExportedNode;
+use PHPStan\ShouldNotHappenException;
+use function array_map;
+use function count;
+
+class ExportedAttributeNode implements ExportedNode, JsonSerializable
+{
+
+	/**
+	 * @param ExportedAttributeArgumentNode[] $args
```
Comment (ondrejmirtes) [inline src/Dependency/ExportedNode/ExportedAttributeNode.php]: `array<int|string, ExportedAttributeArgumentNode>` please

File: src/Dependency/ExportedNode/ExportedAttributeNode.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+		}
+
+		foreach ($this->args as $i => $ourAttribute) {
+			$theirAttribute = $node->args[$i];
+			if (!$ourAttribute->equals($theirAttribute)) {
+				return false;
+			}
+		}
+
+		return $this->name === $node->name;
```
Comment (ondrejmirtes) [inline src/Dependency/ExportedNode/ExportedAttributeNode.php]: The name could be checked sooner

File: src/Dependency/ExportedNodeResolver.php [hunk 4]
```diff
@@ -392,4 +401,32 @@ private function exportClassStatement(Node\Stmt $node, string $fileName, string
...
+	 * @return ExportedAttributeNode[]
+	 */
+	private function exportAttributeNodes(array $attributeGroups): array
+	{
+		$nodes = [];
+		foreach ($attributeGroups as $attributeGroup) {
+			foreach ($attributeGroup->attrs as $attribute) {
+				$args = [];
+				foreach ($attribute->args as $arg) {
+					$args[] = new ExportedAttributeArgumentNode(
```
Comment (ondrejmirtes) [inline src/Dependency/ExportedNodeResolver.php]: This is where the argument should be indexed by `$i` or `$arg–>name`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by ondrejmirtes** [comment]
This is mostly fine :) As for testing - I'd say that static analysis mostly covers that because it's just about passing the correct values into the correct parameters. There's a `ResultCacheEndToEndTest` that tests a real-world project with result cache but it certainly doesn't use attributes because it's an old one.

So please just do the small modifications I asked for in the review, and it's ready to merge :) Thanks.

File: src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php (hunk 1 above)
Comment (olsavmic) [reply src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php]: ~~I don't think so, you can switch named argument and keep the value & index.~~

Ok :)

Comment (olsavmic) [reply src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php]: I removed the src/Dependency/ExportedNode/ExportedAttributeArgumentNode.php entirely as it was storing only the expression value. Let me know if you prefer to keep it

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**